doc/design-component.md

```diff
@@ -0,0 +1,69 @@
+Component Loader Design
+=======================
+
+# Background
+MicroShift is designed to provide a turnkey solution that can readily serve the target workloads 
+with little intervention, probably in an airgapped environment. This requires MicroShift to
+be able to provides the most, if not all, infrastructure components during bootstrapping. 
+
+In its current form, MicroShift has an predefined asset repo and a sequence of 
+loading the manifests to the cluster. Although new assets can be added programmatically, 
+this workload clearly suffers from scalability and flexibilty. 
+
+Specifically, currently the core components are: Service-ca, OpenShift DNS, OpenShift Router, and HostPath provisioner.
+For non-interactive use cases, few of these components are required. While for Data Services oriented workloads 
+may prefer a more reliable and scale-out storage is beyond the capacity of the HostPath volume. 
+
+So it is clear that MicroShift must have a way to customize which components to load during bootstrapping and normal startup.
+
+# Existing Solutions
+
+## Operator
+Operator aims to manage the entire lifecycles of complicated components. Although Operator solves the issue
+of managing inter-dependecy, it is deemed heavier than the MicroShift use case. In many perceived occasions, the MicroShift clusters are deployed just once: they are either stay on offline mode for life (e.g. an Edge device deployed in isolated environment) or destroyed without care (e.g. a developer's test environment). There is no
+need to keep the Operator running to keep the API objects in the desired states. On the other hand, Operators' all-on pattern does not serve MicroShift's small footprint design goal.
+
+## Kubernetes Addon
+Kubernetes Addon mechanism loads cluster wide infrastructure after cluster initialization. Addons such as networking, storage, and management are curated in the Kubernetes repo. Such arrangement, however, does not fit 
+into the MicroShift use case:
+- MicroShift may need more than one components. For instance, Ceph bucket notification, a use case for processing
+Object Store objects on demand, involves Rook/Ceph and Knative (Eventing, Serving, and Networking). 
+- These components are often inter-dependent. Kubernetes Addons cannot ensure the components are created or started in the desired order.
+
+## Helm
+Helm packages multiple components in a chart form and deploy them in a declarative manner. Once deployed, Helm exists till the next it upgrades or removes the chart from the cluster. The non-interactive and low footprint features fit into MicroShift use well. Still the chart must be accessible by the MicroShift cluster when operating
+in an airgapped environment.
+
```

Comment:
I propose to also add `kustomize` into this list.

To me, `kustomize` makes a lot of sense as - unlike Helm - it is K8s native, has mechanisms to apply resources in the right order, and supports remote resources and 'kustomizations', which is nice for a GitOps pattern of operations. It also means we don't need to install an additional dependency.
It's (small) disadvantage vs. Helm is that it doesn't have a default archive format (though it could always be tar-balled, of course).

Maybe we can even support both? E.g. using kustomize built-in and bootstrap a Helm extension with it?

Reply:
I agree, Kustomize yamls sound like the lightest-weight solution we could implement. For helm I wonder what's the memory footprint of the helm processes and images (I have been looking into that, but couldn't find anything), of course those don't run continuously, but when they need to run they will replace other processes's pages or cached pages from memory which is harmful.

Would it make sense to start by something like kustomize, and then if we see gaps, we can look into supporting helm? what do you think?